feature/session/src/main/java/io/github/droidkaigi/confsched2019/session/ui/SessionDetailFragment.kt

```diff
@@ -172,6 +182,50 @@ class SessionDetailFragment : DaggerFragment() {
         }
     }
 
+    private fun drawSessionDescription() {
+        val textView = binding.sessionDescription
+        textView.viewTreeObserver.addOnDrawListener {
+            if (textView.lineCount > 5 && showEllipsis) {
+                val end = textView.layout.getLineStart(5)
+                val ellipsis = getString(R.string.ellipsis_label)
+                val text = buildSpannedString {
```

Comment:
How about a separate method?
This is my code. So sorry 🙇 

```
fun buildDescriptionText(onClickListener: ()->Unit) : CharSequence{
}
```

Reply:
You better understand that day! I will do so. 👍

Reply:
sorry, I don't understand.

why this?

```
fun buildDescriptionText(onClickListener: ()->Unit) : CharSequence{
}
```

Reply:
I will make EllipsisClickableSpan and makeEllipsisColor. Not good?

Reply:
Because when I read the code. 
I don't want to read long Spannables codes. But click logic is important.
So I want to separate code 👀

Reply:
But this is only my preference 🙇

Reply:
ok! 
I'm going to change it as follows, how?
If this is the case, I am parting and I wonder if I can understand what I am doing.

```
val text = buildSpannedString {
   inSpans(makeEllipsisClickableSpan()) {
       append(textView.text.subSequence(0, end - ellipsis.length))
       inSpans(makeEllipsisColor()) { append(ellipsis) }
    }
}

private fun makeEllipsisClickableSpan(): ClickableSpan {
        return object : ClickableSpan() {
            override fun onClick(widget: View) {
                val session = binding.speechSession?.desc
                binding.sessionDescription.text = session
                showEllipsis = !showEllipsis
            }
            override fun updateDrawState(ds: TextPaint) {
                // nothing
            }
        }
}

private fun makeEllipsisColor(): ForegroundColorSpan {
        return  object : ForegroundColorSpan(
            ContextCompat.getColor(requireContext(),
                R.color.colorSecondary
            )
        ) {
            override fun updateDrawState(textPaint: TextPaint) {
                textPaint.color = ContextCompat.getColor(
                    requireContext(),
                    R.color.colorSecondary
                )
                textPaint.isUnderlineText = false
            }
        }
}

```

Reply:
This think is good! Me too.
I separated code https://github.com/DroidKaigi/conference-app-2019/pull/454/commits/8a21b002da26fb3b5c34a537d6f8269a83101b94
Please check.

Reply:
Thanks! I changed some code. 🙏 

* Use View.doOnPreDraw() ktx method for preventing multiple method call
* Use color()  ktx function
* Separate click span as extension function

```kotlin
    private fun drawSessionDescription() {
        val textView = binding.sessionDescription
        textView.doOnPreDraw {
            if (textView.lineCount > 5 && showEllipsis) {
                val end = textView.layout.getLineStart(5)
                val ellipsisText = getString(R.string.ellipsis_label)
                val ellipsisColor = ContextCompat.getColor(requireContext(), R.color.colorSecondary)
                val detailText = textView.text.subSequence(0, end - ellipsisText.length)
                val onClickListener = {
                    val session = binding.speechSession?.desc
                    binding.sessionDescription.text = session
                    showEllipsis = !showEllipsis
                }

                val text = buildSpannedString {
                    clickableSpan(onClickListener, {
                        append(detailText)
                        color(ellipsisColor) {
                            append(ellipsisText)
                        }
                    })
                }
                binding.sessionDescription.setText(text, TextView.BufferType.SPANNABLE)
                binding.sessionDescription.movementMethod = LinkMovementMethod.getInstance()
            }
        }
    }

    private fun SpannableStringBuilder.clickableSpan(
        clickListener: () -> Unit,
        builderAction: SpannableStringBuilder.() -> Unit
    ) {
        inSpans(object : ClickableSpan() {
            override fun onClick(widget: View) {
                clickListener()
            }

            override fun updateDrawState(ds: TextPaint) {
                // nothing
            }
        }, builderAction)
    }
```

Reply:
> ・Use View.doOnPreDraw() ktx method for preventing multiple method call
・Use color() ktx function

👍 I wil do so.
> ・Separate click span as extension function
The code is like kotlin. My code is may not be like kotlin 😢 informative 😄

Reply:
I have a question.
I thought.
1. builder instead of builderAction. Because, I thought what I was doing was not Action.
2. SpannableStringBuilder. () Is a wrong way to write. Why can only write like this? I thought whether it was SpannableStringBuilder () . If you know about Kotlin it might be written like this 😢

Reply:
fixed. Please look!